Initial support for SolidRun CuBox-i devices based on i.MX6 processors (i1, i2, i2eX, and i4Pro).

v-vid opened the pull request:
**NOTE: this commit depends on https://github.com/openwrt/openwrt/pull/904 for the u-boot support as the mainline OpenWrt uboot-imx6 is broken and cannot be compiled at the moment. Current u-boot has been patched for the reviewing purposes only. Once the commit is approved (and #904 is merged) I will resubmit updated PR.**

Specifications

CuBox i1:
- SoC: i.MX6 Solo
- Cores: 1
- Memory Size: 512MB
- GPU: GC880
- Wifi/Bluetooth: Optional
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

CuBox i2 | i2eX:
- SoC: i.MX6 Dual Lite
- Cores: 2
- Memory Size: 1GB
- GPU: GC2000
- Wifi/Bluetooth: Optional
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

CuBox i4Pro | i4x4:
- SoC: i.MX6 Quad
- Cores: 4
- Memory Size: 2/4 GB
- GPU: GC2000
- Wifi/Bluetooth: Build In
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

Built-in u-boot requires SPL (secondary program loader) to be present on the SD-card regardless of the image type which will be loaded.
SPL is generated by the u-boot-mx6cuboxi package which is preselected by the target device and can be found in bin/u-boot-mx6cuboxi directory.

Flashing the SPL:
dd if=/dev/zero of=/dev/mmcblk0 bs=1M count=4
dd if=bin/targets/imx6/generic/u-boot-mx6cuboxi/SPL of=/dev/mmcblk0 bs=1K seek=1

Preparing the firmware on the SD-card:
(echo o; echo n; echo p; echo 1; echo ''; echo ''; echo w) | fdisk /dev/mmcblk0
mkfs.ext4 /dev/mmcblk0p1
mount /dev/mmcblk0p1 /mnt
tar -xzf bin/targets/imx6/generic/openwrt-imx6-device-cubox-i-rootfs.tar.gz -C /mnt/
mkdir -p /mnt/boot
cp bin/targets/imx6/generic/{*-uImage,*.dtb} /mnt/boot/

Generated u-boot.img needs to be placed on the first partition:
cp bin/targets/imx6/generic/u-boot-mx6cuboxi/u-boot.img /mnt/

To boot from the SD card:

setenv image_name 'openwrt-imx6-cubox-i-uImage'; setenv fdt_name 'openwrt-imx6-imx6q-cubox-i.dtb'
setenv kernel_addr_r '0x13000000'; setenv fdt_addr_r 0x12000000; setenv bootargs '$console root=/dev/mmcblk1p1 rw rootwait'
mmc dev 0; ext4load mmc 0:1 $kernel_addr_r /boot/$image_name; ext4load mmc 0:1 $fdt_addr_r /boot/$fdt_name; bootz $kernel_addr_r - $fdt_addr_r

Currently imx6dl-cubox-i.dtb (Dual Lite) and imx6q-cubox-i.dtb (Quad) device trees are available.

Tested on i4Pro, wireless and bluetooth are broken ATM. According to SolidRun forums, BCM4329/BCM4330 firmware is used which works fine on older kernel.

Signed-off-by: Vladimir Vid <vladimir.vid@sartura.hr>

tmn505 commented:
Hi Vladimir.
Some of the CuBox don't have serial console port (Micro USB to RS-232, see: https://wiki.solid-run.com/doku.php?id=products:imx6:cubox-i) and on my CuBoxTV (I4P TV-300-D) I could not interact with U-Boot shell. Maybe instead of altering environment in shell, booting using boot script (boot.scr) would be better? Also using U-Boot autodetection of necessary dtb file (fdt) would widen the spectrum of supported devices (HummingBoard). This is the content of boot.scr i used for booting:
````
setenv bootargs "${console} root=/dev/mmcblk1p1 rw rootwait"

load ${devtype} ${devnum}:${bootpart} ${kernel_addr_r} /boot/openwrt-imx6-cubox-i-uImage
load ${devtype} ${devnum}:${bootpart} ${fdt_addr_r} /boot/openwrt-imx6-${fdtfile}

bootz ${kernel_addr_r} - ${fdt_addr_r}
````
The second issue is this file: `target/linux/imx6/base-files/etc/board.d/01_console`
Why did You decide to alter the line with `tty1` instead of adding it to `target/linux/imx6/base-files/etc/inittab` in the source tree?  Is it problematic for other imx6 devices? With that change one can't access console on HDMI output (for HDMI port to work You'll need this: https://patchwork.ozlabs.org/patch/893585).

For the front led You can pull this https://github.com/chunkeey/apm82181-lede/commit/79ecc475887273d9652c482f1402fc9b3711de2d from Christian repo.

Everything else seems to work fine. Thanks.

bobafetthotmail commented:
On 07/05/2018 22:01, tmn505 wrote:
>
> Hi Vladimir.
> Some of the CuBox don't have serial console port (Micro USB to RS-232, 
> see: https://wiki.solid-run.com/doku.php?id=products:imx6:cubox-i) and 
> on my CuBoxTV (I4P TV-300-D) I could not interact with U-Boot shell.
>
>

Looking at their connector schematics [1] it seems they are using a 
common  FT230x usb-to-uart bridge, which you can also find easily in 
usb-uart dongles. You would only need to find where the RX and TX lines 
are to connect to it with such external dongle.

Just saying it for those interested, I don't think you should do that if 
you can just use the bootscript or whatever other method.


1. 
http://download.solid-run.com/pub/solidrun/cubox-i/Hardware/Schematics-rev-1.1/CuBox-i-Lower.pdf

-Alberto

v-vid commented:
Hi tmn505,

thanks for the feedback. I'm still working on the boot script, HDMI and wireless support, which I was hopping to finish by the time #904 was merged.

As for the console part, I assumed tty1 was used by the other devices so I decided to replace it only for the Cubox in order to avoid breaking something, but perhaps that was not the best idea.

I will fix these and add led support in the updated PR.

Commit:
Initial support for SolidRun CuBox-i devices based on i.MX6 processors (i1, i2, i2eX, and i4Pro).

- Specifications -

CuBox i1:
- SoC: i.MX6 Solo
- Cores: 1
- Memory Size: 512MB
- GPU: GC880
- Wifi/Bluetooth: Optional
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

CuBox i2 | i2eX:
- SoC: i.MX6 Dual Lite
- Cores: 2
- Memory Size: 1GB
- GPU: GC2000
- Wifi/Bluetooth: Optional
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

CuBox i4Pro | i4x4:
- SoC: i.MX6 Quad
- Cores: 4
- Memory Size: 2/4 GB
- GPU: GC2000
- Wifi/Bluetooth: Build In
- USB 2.0 ports: 2
- Ethernet: 10/100/1000 Mbps

Built-in u-boot requires SPL (secondary program loader) to be present on the SD-card regardless of the image type which will be loaded.
SPL is generated by the u-boot-mx6cuboxi package which is preselected by the target device and can be found in bin/u-boot-mx6cuboxi directory.

Flashing the SPL:
dd if=/dev/zero of=/dev/mmcblk0 bs=1M count=4
dd if=bin/targets/imx6/generic/u-boot-mx6cuboxi/SPL of=/dev/mmcblk0 bs=1K seek=1

Preparing the firmware on the SD-card:
(echo o; echo n; echo p; echo 1; echo ''; echo ''; echo w) | fdisk /dev/mmcblk0
mkfs.ext4 /dev/mmcblk0p1
mount /dev/mmcblk0p1 /mnt
tar -xzf bin/targets/imx6/generic/openwrt-imx6-device-cubox-i-rootfs.tar.gz -C /mnt/
mkdir -p /mnt/boot
cp bin/targets/imx6/generic/{*-uImage,*.dtb,*.scr} /mnt/boot/

Generated u-boot.img needs to be placed on the first partition:
cp bin/targets/imx6/generic/u-boot-mx6cuboxi/u-boot.img /mnt/

To boot from the SD card:

Boot script which sets mmc/dtb parameters and boots the board is automatically sourced.
If this does not work for any reason:
mmc dev 0; load mmc 0:1 $scriptaddr boot/boot.scr; source $scriptaddr

Currently imx6dl-cubox-i.dtb (Dual Lite) and imx6q-cubox-i.dtb (Quad) device trees are available.

Tested on i4Pro, MMC, USB (+ HiD), HDMI and ethernet ports are working.
Wireless and bluetooth are broken ATM. According to SolidRun forums, BCM4329/BCM4330 firmware is used which works fine on older kernels.

Signed-off-by: Vladimir Vid <vladimir.vid@sartura.hr>

v-vid commented:
Hi @tmn505,
I went with your suggestion and added bootscript which chooses dtb file depending on the _board_name_ and _board_rev_ parameters. Currently this is limited only to Cubox Dual Lite and Quad since their dts's are available in the mainline, but I hope we can add the other devices as well.

_kmod-drm-imx_ and _kmod-drm-imx-hdmi_ packages solved the HDMI issue, it is now accessable in both u-boot and userspace, 01_console was removed.

_kmod-usb-hid_ was also added to support the USB keyboard so the device can be fully accessible without serial port and/or ssh.

I did not have time to fully test the front led, so I'm planning to add that later along with the wireless support (still currently working on this, but I've made some progress).

Could you retest this again if you find some time?

hauke commented:
Thank you for the pull request, it was applied to master.
